### DISCUSSION TITLE
Recommended method of importing components from other Nx libs in standalone environment

### DISCUSSION THREAD

**Discussion by danhobbs75** [nrwl/nx#17665] [General]
Hello. I've recently converted my Angular app (single app with lots of Nx libs) to use standalone components, directives and pipes throughout. I'm pretty happy with this, but I've not found any good guidance on how to import components from other libs now that they're not grouped together in modules.

I may be overcomplicating this, but I am concerned I'm making it difficult for Nx when it decides what goes in each bundle. For instance, I now don't have my components showing up in barrel index.ts files, and the language service is automatically importing using a relative path. It feels wrong to me. But is it? :-)

As an example, previously I would see

`import { ExampleComponent } from '@app/libraryA';`

but now I'm seeing

`import { ExampleComponent } from 'libs/libraryA/src/lib/example.component';`

I'd love to know if anyone has seen a good discussion of this anywhere. Maybe this is the start of it if there is someone who understands this stuff much better than me and wants to explain how they think it should be...

**Comment by danielbrazg** [comment]
Hi! I also want to know the difference and what is considered the right way to import components from libraries.
in my case, it affects the build process.

a detailed explanation of my question:

example of my project structure: 
```
apps/client-a
apps/client-b 
lib/client-components/icon-component (not using external dependencies)
lib/client-components/calendar-component (using moment)
```

currently, I use NX default behavior so I import in both apps the components like this:
client-a: `import {IconComponent, CalendarComponent} from '@my-project/client-components';`
client-b: `import {IconComponent} from '@my-project/client-components';`

in this example, I get an error when trying to build client-b:
`./node_modules/moment/index.js - Error: Module not found: Error: Can't resolve 'moment' in 'my-project/node_modules/moment'`

note: this is working when I remove the export CalendarComponent  line from the lib index.ts file

I succeeded in solving it by changing the import in client-b to:
`import {IconComponent} from 'libs/client-components/src/lib/icon/icon.component';
`but it seems like it's the wrong solution.

so should be the right way to do that?

**Reply by danhobbs75** [reply]
I've made some progress with my setup, and I was on the same track you were originally where I now have all my libs with all the components, directives and pipes exported from lib barrel files.

They do have external dependencies in many cases, but they are building correctly. I think you were on the right track originally importing from libraries, and I agree that it's wrong to import using a relative path.

The build problem with moment is the issue and I would look to solve that if I were you. I don't use moment myself but it often comes up in tree-shaking discussions and I've seen articles where much smaller third party libraries do the job of a part of moment and some people are replacing moment with those. Might be worth a try...?

**Reply by danielbrazg** [reply]
I agree with you. it was just for the example. but actually, in my app. I don't use Moment directly, but another package that I use uses Moment so I don't have much control over it.
But what I'm trying to understand, is why different component dependencies are still shown in the build process. if this is just a Moment issue, how can I fix it so the build process will be done correctly?

**Reply by danhobbs75** [reply]
Ah, I understand. I don't think you're doing anything wrong at all, so I can't add anything extra now. Maybe worth checking that the package you're using has its dependencies defined correctly and that Moment is indeed in your dev dependencies (and you can see it install in node_modules). Perhaps try importing Moment directly in your calendar component and see if that helps bring it in at build time...I know you shouldn't have to, but maybe being explicit in that import will bring you closer to the real reason by giving you clues.
